src/Operation/MapReduce.php

```diff
@@ -0,0 +1,319 @@
+        $cursor = $server->executeCommand($this->databaseName, $this->createCommand($server), $readPreference);
+
+        $getIterator = $this->createGetIteratorCallable($result, $server);
+        $counts = isset($result->counts) ? (array) $result->counts : null;
+        $executionTimeMS = isset($result->timeMillis) ? $result->timeMillis : null;
+
+        return new MapReduceResult($getIterator, $counts, $executionTimeMS);
+    }
+
+    /**
+     * Create the mapReduce command.
+     *
+     * @param Server $server
+     * @return Command
+     */
+    private function createCommand(Server $server)
+    {
+        $cmd = [
+            'mapReduce' => $this->collectionName,
+            'map' => $this->map,
+            'reduce' => $this->reduce,
+            'out' => $this->out,
+        ];
+
+        foreach (['finalize', 'jsMode', 'limit', 'maxTimeMS', 'readConcern', 'verbose', 'writeConcern'] as $option) {
+            if (isset($this->options[$option])) {
+                $cmd[$option] = $this->options[$option];
+            }
+        }
+
+        foreach (['collation', 'query', 'scope', 'sort'] as $option) {
+            if (isset($this->options[$option])) {
+                $cmd[$option] = (object) $this->options[$option];
+            }
+        }
+
+        if (isset($this->options['bypassDocumentValidation']) && \MongoDB\server_supports_feature($server, self::$wireVersionForDocumentLevelValidation)) {
+            $cmd['bypassDocumentValidation'] = $this->options['bypassDocumentValidation'];
+        }
+
+        return new Command($cmd);
+    }
+
+    /**
+     * Creates a callable for MapReduceResult::getIterator().
+     *
+     * @param stdClass $result
+     * @param Server   $server
+     * @return callable
+     * @throws UnexpectedValueException if the command response was malformed
+     */
+    private function createGetIteratorCallable(stdClass $result, Server $server)
+    {
+        // Inline results can be wrapped with an ArrayIterator
+        if (isset($result->results) && is_array($result->results)) {
+            $results = $result->results;
+
+            return function() use ($results) {
+                if (isset($this->typeMap)) {
+                    return new TypeMapArrayIterator($result->results, $this->typeMap);
+                }
+
+                return new ArrayIterator($reesult->results);
+            };
+        }
+
+        if (isset($result->result) && (is_string($result->result) || is_object($result->result))) {
+            $options = isset($this->typeMap) ? ['typeMap' => $this->typeMap] : [];
+
+            $find = is_string($result->result)
+                ? new Find($this->databaseName, $result->result, $options)
+                : new Find($result->result->db, $result->result->collection, $options);
+
+            return function() use ($find, $server) {
+                return $find->execute($server);
```

Comment:
Nice idea for returning results in case an output collection was used. I might have to remember this for ODM as it would be useful when using an `$out` stage in aggregation builder.